syntax/parser.go

```diff
@@ -214,80 +218,99 @@ func (p *parser) noteCaptureName(name string, pos int) {
 		p.capnames = make(map[string]int)
 	}
 
+	if p.useMaintainCaptureOrder() {
+		if p.capnamenums == nil {
+			p.capnamenums = make(map[string]int)
+		}
+
+		if _, ok := p.capnamenums[name]; !ok {
+			p.capnamenums[name] = p.autocap
+			p.noteCaptureSlot(p.consumeAutocap(), pos)
+		}
+	}
+
 	if _, ok := p.capnames[name]; !ok {
 		p.capnames[name] = pos
 		p.capnamelist = append(p.capnamelist, name)
 	}
 }
 
 func (p *parser) assignNameSlots() {
-	if p.capnames != nil {
-		for _, name := range p.capnamelist {
-			for p.isCaptureSlot(p.autocap) {
+	if p.useMaintainCaptureOrder() {
+		p.capnames = p.capnamenums
+		// Prepend `0` to capnamelist if it's not set (MaintainCaptureOrder was enabled inline)
+		if len(p.capnamelist) == 0 || p.capnamelist[0] != `0` {
+			p.capnamelist = append([]string{fmt.Sprint(0)}, p.capnamelist...)
+		}
```

Comment:
can this just return and not use the else?  should clean up the diff quite a bit.